File: server/test_replay_parity.py
"""
Python replay motoru parite testi.

server/replay_fixtures.json içindeki transkriptler istemci mantığıyla
(scripts/gen-replay-fixtures.mjs) üretilmiştir. Python replay_game HER
transkript için AYNI skor ve en büyük kareyi vermelidir.

Çalıştır:  python3 server/test_replay_parity.py
"""
import json
import os
import sys

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
from replay import replay_game  # noqa: E402

HERE = os.path.dirname(os.path.abspath(__file__))
FIXTURES = os.path.join(HERE, "replay_fixtures.json")


def main():
    with open(FIXTURES, encoding="utf-8") as f:
        fixtures = json.load(f)

    fails = 0
    longest = 0
    for i, fx in enumerate(fixtures):
        r = replay_game(fx["seed"], fx["moves"], fx["size"])
        longest = max(longest, len(fx["moves"]))
        if not r["valid"]:
            print(f"  [{i}] GECERSIZ (beklenmedik): seed={fx['seed']}")
            fails += 1
            continue
        if r["score"] != fx["score"]:
            print(
                f"  [{i}] SKOR FARKI: py={r['score']} != js={fx['score']} "
                f"(seed={fx['seed']} size={fx['size']} moves={len(fx['moves'])})"
            )
            fails += 1
        if r["maxTile"] != fx["maxTile"]:
            print(
                f"  [{i}] MAXTILE FARKI: py={r['maxTile']} != js={fx['maxTile']}"
            )
            fails += 1

    total = len(fixtures)
    if fails:
        print(f"\n[X] PARITE BASARISIZ: {fails} uyusmazlik / {total} fixture")
        sys.exit(1)
    print(
        f"[OK] PARITE TAM: {total} fixture, en uzun oyun {longest} hamle - "
        f"Python replay = istemci mantigi (birebir)"
    )


if __name__ == "__main__":
    main()

File: server/test_replay_parity.ts
// ============================================================
//  TypeScript replay motoru parite testi (server/replay.ts).
//
//  server/replay_fixtures.json içindeki transkriptler istemci mantığıyla
//  (scripts/gen-replay-fixtures.mjs) üretilmiştir. Sunucu replayGame HER
//  transkript için AYNI skor ve en büyük kareyi vermelidir — aksi hâlde
//  NestJS backend'i meşru oyunları reddeder (yanlış pozitif).
//
//  Altyapısız çalıştır (Node ≥ 22.6 --experimental-strip-types, ≥ 23.6 doğal):
//    node server/test_replay_parity.ts
//
//  Python eşi: python3 server/test_replay_parity.py (ikisi de geçmeli).
// ============================================================

import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { dirname, join } from 'node:path';
import { replayGame } from './replay.ts';

interface Fixture {
  seed: number;
  size: number;
  moves: string;
  score: number;
  maxTile: number;
}

const HERE = dirname(fileURLToPath(import.meta.url));
const FIXTURES = join(HERE, 'replay_fixtures.json');

function main(): void {
  const fixtures: Fixture[] = JSON.parse(readFileSync(FIXTURES, 'utf-8'));

  let fails = 0;
  let longest = 0;
  for (let i = 0; i < fixtures.length; i++) {
    const fx = fixtures[i];
    const r = replayGame(fx.seed, fx.moves, fx.size);
    longest = Math.max(longest, fx.moves.length);
    if (!r.valid) {
      console.log(`  [${i}] GECERSIZ (beklenmedik): seed=${fx.seed}`);
      fails++;
      continue;
    }
    if (r.score !== fx.score) {
      console.log(
        `  [${i}] SKOR FARKI: ts=${r.score} != js=${fx.score} ` +
          `(seed=${fx.seed} size=${fx.size} moves=${fx.moves.length})`,
      );
      fails++;
    }
    if (r.maxTile !== fx.maxTile) {
      console.log(`  [${i}] MAXTILE FARKI: ts=${r.maxTile} != js=${fx.maxTile}`);
      fails++;
    }
  }

  const total = fixtures.length;
  if (fails) {
    console.log(`\n[X] PARITE BASARISIZ: ${fails} uyusmazlik / ${total} fixture`);
    process.exit(1);
  }
  console.log(
    `[OK] PARITE TAM: ${total} fixture, en uzun oyun ${longest} hamle - ` +
      `TypeScript (sunucu) replay = istemci mantigi (birebir)`,
  );
}

main();
